IdentityCore/src/throttling/cache/MSIDLRUCache.m

```diff
@@ -0,0 +1,432 @@
+        }
+        return NO;
+    }
+    
+    MSIDLRUCacheNode *node = [self.container objectForKey:signature];
+    MSIDLRUCacheNode *prevNode = [self.container objectForKey:node.prevSignature];
+    MSIDLRUCacheNode *nextNode = [self.container objectForKey:node.nextSignature];
+
+    prevNode.nextSignature = node.nextSignature;
+    nextNode.prevSignature = node.prevSignature;
+
+    [self.container setObject:prevNode forKey:node.prevSignature];
+    [self.container setObject:nextNode forKey:node.nextSignature];
+    
+    [self.container removeObjectForKey:signature];
+    return YES;
+}
+
+//retrieve cache record from the corresponding node, and move the node to the front of LRU cache.
+- (id)updateAndReturnCacheRecord:(id)key
+                           error:(NSError **)error
+{
+    __block id cacheRecord;
+    __block NSError *subError = nil;
+    
+    dispatch_sync(self.synchronizationQueue, ^{
+        if (![self.key_signature_map objectForKey:key])
+        {
+            subError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"MSIDLRUCache Error: Unable to find valid signature for the input key during retrieval", nil, nil, nil, nil, nil, YES);
+        }
+        
+        cacheRecord = [self updateAndReturnCacheRecordImpl:[self.key_signature_map objectForKey:key]
+                                                     error:&subError];
+    });
+    if (subError)
+    {
+        cacheRecord = nil;
+    }
+    
+    if (error)
+    {
+        *error = subError;
+    }
+
+    return cacheRecord;
+}
+
+- (id)updateAndReturnCacheRecordImpl:(NSString *)signature
+                               error:(NSError **)error
+{
+    if (![self.container objectForKey:signature])
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"MSIDLRUCache Error: Unable to find valid node for the input signature during retrieval", nil, nil, nil, nil, nil, YES);
+        }
+        return nil;
+    }
+    
+    //retrieve node
+    MSIDLRUCacheNode *node = [self.container objectForKey:signature];
+    
+    //remove from current cache slot
+    [self removeFromCacheImpl:signature
+                        error:error];
+    //move to front
+    [self addToFrontImpl:node];
+    
+    return node.cacheRecord;
+}
+
+- (void)addToFront:(MSIDLRUCacheNode *)node
+{
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        [self addToFrontImpl:node];
+    });
+}
+
+- (void)addToFrontImpl:(MSIDLRUCacheNode *)node
+{
+    NSString *currentHeadSignature = self.head.nextSignature; //node currently pointed by the head
+    MSIDLRUCacheNode *currentHeadNode = [self.container objectForKey:currentHeadSignature];
+    
+    /**
+    BEFORE:
+     A->B->C
+     A<-B<-C
+    AFTER:
+     A->C
+     A<-C
+     */
+    currentHeadNode.prevSignature = [node.signature mutableCopy];
```

Comment:
any reason why we need to do `mutableCopy`?

Reply:
input param is NSString, and this variable is NSMutable string.